js/ui/iconGrid.js

```diff
@@ -477,6 +477,11 @@ const IconGrid = new Lang.Class({
         this.saturation = new Shell.GridDesaturateEffect({ factor: 0,
                                                            enabled: false });
         this.actor.add_effect(this.saturation);
+
+        Main.layoutManager.connect('monitors-changed', Lang.bind(this, function() {
+            this.compositeMode = Main.layoutManager.primaryMonitor &&
+                                 Main.layoutManager.primaryMonitor.width < 800;
```

Comment:
You could factor this into an `updateCompositeMode()` function and also call it once after connecting to the signal. That should make it possible to avoid calling it every size allocate.